lib/mongodb.js

```diff
@@ -2023,28 +2062,57 @@ function visitAllProperties(data, modelCtorOrDef, visitor) {
         visitAllProperties(value, def.type.definition, visitor);
       }
     } else {
-      visitor(value, def, (newValue) => { data[p] = newValue; });
+      visitor(modelCtor, value, def, (newValue) => { data[p] = newValue; });
     }
     continue;
   }
 }
 
 /**
- *
- * @param {*} propValue Property value to coerce into a Decimal128 value
- * @param {*} propDef Property definition to check if property is MongoDB
- * Decimal128 type
+ * @param {*} modelCtor Model constructor
+ * @param {*} propValue Property value to coerce into special types supported by the connector
+ * @param {*} propDef Property definition to check if property is for MongoDB
  */
-function coerceDecimalProperty(propValue, propDef, setValue) {
-  let updatedValue;
-  if (hasDataType('decimal128', propDef)) {
-    if (Array.isArray(propValue)) {
-      updatedValue = propValue.map(val => Decimal128.fromString(val));
-      return setValue(updatedValue);
-    } else {
-      updatedValue = Decimal128.fromString(propValue);
-      return setValue(updatedValue);
+function coerceProperties(modelCtor, propValue, propDef, setValue) {
```

Comment:
I find the name misleading. It says "coerce properties" (plural name), but it's coercing a single property value AFAICT.

Proposed name:
 
```suggestion
function coercePropertyValue(modelCtor, propValue, propDef, setValue) {
```